crypto/fipsmodule/rand/new_rand.c

```diff
@@ -0,0 +1,290 @@
+        personalization_string_len) != 1) {
+    abort();
+  }
+
+  state->generate_calls_since_seed = 0;
+
+  OPENSSL_cleanse(seed, CTR_DRBG_ENTROPY_LEN);
+  OPENSSL_cleanse(personalization_string, CTR_DRBG_ENTROPY_LEN);
+}
+
+// rand_state_initialize initializes the thread-local state |state|. In
+// particular initializes the CTR-DRBG state with the initial seed material.
+static void rand_state_initialize(struct rand_thread_local_state *state) {
+
+  get_entropy_source(&state->entropy_source);
+
+  uint8_t seed[CTR_DRBG_ENTROPY_LEN];
+  uint8_t personalization_string[CTR_DRBG_ENTROPY_LEN];
+  size_t personalization_string_len = 0;
+  rand_get_ctr_drbg_seed_entropy(&state->entropy_source, seed,
+    personalization_string, &personalization_string_len);
+
+  assert(personalization_string_len == 0 ||
+         personalization_string_len == CTR_DRBG_ENTROPY_LEN);
+
+  if (!CTR_DRBG_init(&state->drbg, seed, personalization_string,
+        personalization_string_len)) {
+    abort();
+  }
+
+  state->generate_calls_since_seed = 0;
+
+  OPENSSL_cleanse(seed, CTR_DRBG_ENTROPY_LEN);
+  OPENSSL_cleanse(personalization_string, CTR_DRBG_ENTROPY_LEN);
+}
+
+// RAND_bytes_core generates |out_len| bytes of randomness and puts them in
+// |out|. The CTR-DRBG state |state| is managed to ensure uniqueness and usage
+// requirements are met.
+//
+// The argument |use_user_pred_resistance| must be either
+// |RAND_USE_USER_PRED_RESISTANCE| or |RAND_NO_USER_PRED_RESISTANCE|. The former
+// cause the content of |user_pred_resistance| to be mixed in as prediction
+// resistance. The latter ensures that |user_pred_resistance| is not used.
+static void RAND_bytes_core(
+  struct rand_thread_local_state *state,
+  uint8_t *out, size_t out_len,
+  const uint8_t user_pred_resistance[RAND_PRED_RESISTANCE_LEN],
+  int use_user_pred_resistance) {
+
+  // Ensure CTR-DRBG state is unique.
+  if (rand_ensure_ctr_drbg_uniquness(state, out_len) != 1) {
+    rand_ctr_drbg_reseed(state);
+  }
+
+  // If a prediction resistance source is available, use it.
+  // Prediction resistance is only used on first invocation of the CTR-DRBG,
+  // ensuring that its state is randomized before using output.
+  size_t first_pred_resistance_len = 0;
+  uint8_t pred_resistance[RAND_PRED_RESISTANCE_LEN] = {0};
+  rand_maybe_get_ctr_drbg_pred_resistance(state, pred_resistance,
+    &first_pred_resistance_len);
+
+  // If caller input user-controlled prediction resistance, use it.
+  if (use_user_pred_resistance == RAND_USE_USER_PRED_RESISTANCE) {
+    for (size_t i = 0; i < RAND_PRED_RESISTANCE_LEN; i++) {
+      pred_resistance[i] ^= user_pred_resistance[i];
+    }
+    first_pred_resistance_len = RAND_PRED_RESISTANCE_LEN;
+  }
+
+  assert(first_pred_resistance_len == 0 ||
+         first_pred_resistance_len == RAND_PRED_RESISTANCE_LEN);
+
+  // Iterate CTR-DRBG generate until we generated |out_len| bytes of randomness.
+  while (out_len > 0) {
```

Comment:
Can this be simplified? From my reading of CTR_DRBG_generate it either generates the requested amount of bytes or returns 0 for error, and then you abort down on L216.

Reply:
I don't see how because the generate function can output at most `CTR_DRBG_MAX_GENERATE_LENGTH` per invocation. So, we must loop over invocations to satisfy `out_len` that could be higher.

We can simplify if we restrict `RAND_bytes()` and friends to only allow `out_len <= CTR_DRBG_MAX_GENERATE_LENGTH`.